### DISCUSSION TITLE
Refresh of a dependency in error state

### DISCUSSION THREAD

**Discussion by roubachof** [rrousselGit/riverpod#1890] [Q&A]
I am doing a complete "real life app" PoC to test deeply riverpod possibilities, I'm super happy with the result so far but I am now at the error scenario step, and I have met this case.

`LampListProvider` depends on `ZoneStreetLampsProvider( zone: someZone )`, both are generated async notifiers, the second one is a family.

`LampListProvider` is loading, but `ZoneStreetLampsProvider` is failing.
So they both go to error state (i guess).

Now I wan't to reload `LampListProvider`, i do it by calling `ref.refresh(lampListProvider)`.
The issue I have is that is doesn't seem to trigger a rebuild of the dependency (zoneStreetLampsProvider)...
I though that since the dependency is in error the refresh of the dependent provider would trigger the rebuild of its dependency...

Should I do something else, maybe a function to call or a use case to implement to support this scenario?

**Comment by rrousselGit** [comment]
You're refreshing the wrong provider.

You should refresh `ZoneStreetLampsProvider` instead. Since LampListProvider depends on it, it will also refresh when ZoneStreetLampsProvider refreshes

**Reply by roubachof** [reply]
Indeed I could do that...
But since my widget is using `LampListProvider`, I don't know which provider failed.
And maybe I will be refreshing a provider that will induce a remote call whereas it's just the top level provider that doing some computation on the result of the remote call that is failing...
Also, with a complete Provider architecture, there could be a lot of dependencies, so it would be tedious to trace all the providers...

Don't you think it would make sense that a dependency should be refreshed automatically in this case if it's in error state?

**Reply by rrousselGit** [reply]
I don't think that makes sense, no. Such behavior would be counter-intuitive to how Riverpod works as a whole

Riverpod is all about a graph that build from top to bottom. If somehow refreshing a bottom node can now refresh higher-level nodes too, that makes the logic backward and causes a bunch of problem.

**Reply by roubachof** [reply]
I think there is a misunderstanding cause that's exactly my point :)
the `LampListProvider` is my top node used by my consumer widget, the one visible by my UI hierarchy.
So when I refresh it, I thought it would refresh all the bottom node accordingly (its dependencies accordingly).

**Reply by rrousselGit** [reply]
It isn't your actual top node though. ZoneStreetLampsProvider is higher in the graph than LampListProvider

**Reply by roubachof** [reply]
Ok we both see a tree, I just see it upside down :)
In my opinion if the provider used by a consumer is not the leading node, it will cause some challenges that I am actually facing.
For example the `LampListProvider` depends on 2 others providers, transitively from `ZoneStreetLampsProvider`.
For doing a refresh when I am in error state, it means that I have to trace all those dependency and invalidate each root.
It's feasible, but not a smooth experience.

**Reply by roubachof** [reply]
What I'm trying to say, it that if the refresh doesn't propagate to its dependencies, it breaks encapsulation and the interface semantics. A consumer shouldn't have to know about the dependencies of the interface it's using.
I made a simple schema to try to understand my point.

![riverpod_refresh](https://user-images.githubusercontent.com/596903/200813352-07ecb8ba-4c2b-45ca-8e44-542b0c0287b2.jpeg)

such concerns have also been raised here: https://github.com/rrousselGit/riverpod/issues/698

**Reply by rrousselGit** [reply]
Just add a "refresh" method on your object that's visible to your consumer and call that.

Your method should be able to refresh the dependencies it needs.

**Reply by roubachof** [reply]
I did try this solution but ended up with future == null assert violations...
Let me refactor a bit the providers, finish the PoC so we can discuss on it.
I'm sure I must be doing something fishy :)
I'm still acclimating with the revolution riverpod brings to state management.
